### PR TITLE
Fix OOM issue when publishing big messages

### PR DESCRIPTION
**Pull request by a5anka**
## Purpose

Fix the OOM issue identified in https://github.com/ballerina-platform/ballerina-message-broker/issues/426.

- Stop writing to Netty channel when the outbound buffer is full.
- Stop autoreading when the flow control limits are hit

## Automation tests
 - Unit tests - 77%
 - Integration tests  - https://github.com/ballerina-platform/ballerina-message-broker/issues/451

### CODE REVIEW

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java [hunk 1]
```diff
@@ -79,6 +80,11 @@
      */
     private static final int BLOCKING_TASK_EXECUTOR_THREADS = 32;
 
+    /**
+     * Socket receive and send buffer size in bytes.
+     */
+    private static final int SOCKET_BUFFER_SIZE = 1048576;
```
Comment (sdkottegoda) [inline modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java]: Will it be better if we can make this configurable?

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java [hunk 2]
```diff
@@ -220,7 +232,7 @@ protected void initChannel(SocketChannel socketChannel) {
                          .addLast(sslHandlerFactory.create())
                          .addLast(new AmqpDecoder(amqMethodRegistryFactory.newInstance()))
                          .addLast(new AmqpEncoder())
-                         .addLast(new AmqpConnectionHandler(configuration, metricManager))
+                         .addLast(new AmqpConnectionHandler(metricManager, amqpChannelFactory))
```
Comment (sdkottegoda) [inline modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java]: amqpChannelFactory is initialized with the metricManager, so do we need to pass the metricManager here again?

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java (hunk 1 above)
Comment (a5anka) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java]: :+1:

File: modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java (hunk 2 above)
Comment (a5anka) [reply modules/broker-amqp/src/main/java/io/ballerina/messaging/broker/amqp/Server.java]: This is done to hide AMQP channel creational dependencies from AmqpConnectionHandler.

**Comment by a5anka** [comment]
@sdkottegoda I added a config for socket buffer size. Can you please review?

**Pull request merged**